File: operon/core/exceptions.py
"""Unified exception classes cho Operon Core nodes.

Cung cấp base exception class với context đầy đủ để debug,
cho phép các node-specific exceptions kế thừa và mở rộng.
"""

from typing import Any, Dict, List, Optional


def _truncate(text: str, max_length: int = 200) -> str:
    """Truncate text và thêm ellipsis nếu vượt quá max_length."""
    if len(text) <= max_length:
        return text
    return text[:max_length] + "..."


class OpError(Exception):
    """Base exception cho tất cả các node errors.

    Chứa context đầy đủ để debug:
    - op_type: Loại node (parser, llm, code, ...)
    - original_error: Exception gốc
    - context: Dict chứa thông tin bổ sung tùy theo node type

    Example:
        raise OpError(
            message="Failed to process",
            op_type="parser",
            original_error=json_error,
            context={"input": text, "format": "json"}
        )
    """

    def __init__(
        self,
        message: str,
        op_type: str,
        original_error: Optional[Exception] = None,
        context: Optional[Dict[str, Any]] = None,
    ):
        self.op_type = op_type
        self.original_error = original_error
        self.context = context or {}

        # Build formatted message
        full_message = self._format_message(message)
        super().__init__(full_message)

    def _format_message(self, message: str) -> str:
        """Format error message với context."""
        parts = [f"[{self.op_type.upper()}] {message}"]

        if self.original_error:
            parts.append(f"  Error: {self.original_error}")

        for key, value in self.context.items():
            if isinstance(value, str):
                value = _truncate(repr(value))
            parts.append(f"  {key}: {value}")

        return "\n".join(parts)


class ParserError(OpError):
    """Exception khi parsing thất bại.

    Context tự động bao gồm:
    - input: Text đầu vào (truncated)
    - format: Loại format (json, xml, yaml)

    Example:
        raise ParserError(
            message="Invalid JSON syntax",
            input_text=raw_text,
            format_type="json",
            original_error=json_decode_error
        )
    """

    def __init__(
        self,
        message: str,
        input_text: str,
        format_type: str,
        original_error: Optional[Exception] = None,
    ):
        self.input_text = input_text
        self.format_type = format_type

        super().__init__(
            message=message,
            op_type="parser",
            original_error=original_error,
            context={"format": format_type, "input": input_text},
        )


class CodeError(OpError):
    """Exception khi user function trong FuncOp thất bại.

    Context tự động bao gồm:
    - function_name: Tên function
    - source: Source code (truncated)
    - inputs: Dict các input values (truncated)

    Example:
        raise CodeError(
            message="Function raised an exception",
            function_name="calculate_total",
            source="def calculate_total(x, y): ...",
            inputs={"x": 1, "y": 2},
            original_error=division_error
        )
    """

    def __init__(
        self,
        message: str,
        function_name: str,
        source: str,
        inputs: Dict[str, Any],
        original_error: Optional[Exception] = None,
    ):
        self.function_name = function_name
        self.source = source
        self.inputs = inputs

        super().__init__(
            message=message,
            op_type="code",
            original_error=original_error,
            context={
                "function_name": function_name,
                "source": _truncate(source, 300),
                "inputs": {k: _truncate(repr(v), 100) for k, v in inputs.items()},
            },
        )


class BranchError(OpError):
    """Exception khi đánh giá điều kiện branch thất bại.

    Context tự động bao gồm:
    - condition: Biểu thức điều kiện
    - inputs: Dict các input values (truncated)
    - candidates: List các target node có thể

    Example:
        raise BranchError(
            message="Condition evaluation failed",
            condition="score >= 90",
            inputs={"score": "invalid"},
            candidates=["excellent", "good", "fail"],
            original_error=type_error
        )
    """

    def __init__(
        self,
        message: str,
        condition: str,
        inputs: Dict[str, Any],
        candidates: List[str],
        original_error: Optional[Exception] = None,
    ):
        self.condition = condition
        self.inputs = inputs
        self.candidates = candidates

        super().__init__(
            message=message,
            op_type="branch",
            original_error=original_error,
            context={
                "condition": condition,
                "inputs": {k: _truncate(repr(v), 100) for k, v in inputs.items()},
                "candidates": candidates,
            },
        )


class ConditionError(OpError):
    """Exception when WhileOp ``until`` condition fails.

    Context includes:
    - condition: The ``until`` expression string.
    - phase: 'compile' or 'eval'.
    - iteration: Current iteration number (if available).
    - inputs: Dict of input values (truncated).

    Example:
        raise ConditionError(
            message="Stop condition evaluation failed",
            condition="counter >= 5",
            inputs={"count": 3},
            iteration=3,
            phase="eval",
            original_error=name_error
        )
    """

    def __init__(
        self,
        message: str,
        condition: str,
        inputs: Optional[Dict[str, Any]] = None,
        iteration: Optional[int] = None,
        phase: str = "eval",
        original_error: Optional[Exception] = None,
    ):
        self.condition = condition
        self.inputs = inputs or {}
        self.iteration = iteration
        self.phase = phase

        context = {"condition": condition, "phase": phase}
        if inputs:
            context["inputs"] = {k: _truncate(repr(v), 100) for k, v in inputs.items()}
        if iteration is not None:
            context["iteration"] = iteration

        super().__init__(
            message=message, op_type="while", original_error=original_error, context=context
        )


class IterationError(OpError):
    """Exception khi một iteration trong ForLoop/MapOp thất bại.

    Context tự động bao gồm:
    - iteration_index: Vị trí iteration bị lỗi
    - total_iterations: Tổng số iterations
    - loop_data: Data của iteration (truncated)

    Example:
        raise IterationError(
            message="Iteration 2 failed",
            iteration_index=2,
            loop_data={"item": {"id": 3}},
            total_iterations=10,
            op_type="for",
            original_error=key_error
        )
    """

    def __init__(
        self,
        message: str,
        iteration_index: int,
        loop_data: Dict[str, Any],
        total_iterations: int,
        op_type: str = "for",
        original_error: Optional[Exception] = None,
    ):
        self.iteration_index = iteration_index
        self.loop_data = loop_data
        self.total_iterations = total_iterations

        super().__init__(
            message=message,
            op_type=op_type,
            original_error=original_error,
            context={
                "iteration_index": f"{iteration_index}/{total_iterations}",
                "loop_data": {k: _truncate(repr(v), 100) for k, v in loop_data.items()},
            },
        )


# =============================================================================
# Provider Node Errors (operon-providers)
# =============================================================================


class PromptError(OpError):
    """Exception khi PromptOp format template thất bại.

    Context tự động bao gồm:
    - template_type: Loại template (str/dict/list)
    - template: Preview của template (truncated)
    - missing_vars: List các biến bị thiếu (nếu có)

    Example:
        raise PromptError(
            message="Missing template variable",
            template="Hello {name}, your order {order_id}",
            missing_vars=["order_id"],
            original_error=key_error
        )
    """

    def __init__(
        self,
        message: str,
        template: Any,
        missing_vars: Optional[List[str]] = None,
        original_error: Optional[Exception] = None,
    ):
        self.template = template
        self.missing_vars = missing_vars or []

        context = {
            "template_type": type(template).__name__,
            "template": _truncate(repr(template), 300),
        }
        if missing_vars:
            context["missing_vars"] = missing_vars

        super().__init__(
            message=message, op_type="prompt", original_error=original_error, context=context
        )


class EmbeddingError(OpError):
    """Exception khi EmbeddingOp thất bại.

    Context tự động bao gồm:
    - resource: Model/resource key
    - text_count: Số lượng texts cần embed

    Example:
        raise EmbeddingError(
            message="Embedding backend failed",
            resource="bge-m3",
            text_count=100,
            original_error=connection_error
        )
    """

    def __init__(
        self,
        message: str,
        resource: str,
        text_count: int,
        original_error: Optional[Exception] = None,
    ):
        self.resource = resource
        self.text_count = text_count

        super().__init__(
            message=message,
            op_type="embedding",
            original_error=original_error,
            context={"resource": resource, "text_count": text_count},
        )


class RerankError(OpError):
    """Exception khi RerankOp thất bại.

    Context tự động bao gồm:
    - resource: Model/resource key
    - query: Query string (truncated)
    - document_count: Số lượng documents

    Example:
        raise RerankError(
            message="Invalid document type",
            resource="bge-m3",
            query="search query",
            document_count=50,
            original_error=type_error
        )
    """

    def __init__(
        self,
        message: str,
        resource: str,
        query: str,
        document_count: int,
        original_error: Optional[Exception] = None,
    ):
        self.resource = resource
        self.query = query
        self.document_count = document_count

        super().__init__(
            message=message,
            op_type="rerank",
            original_error=original_error,
            context={
                "resource": resource,
                "query": _truncate(query, 100),
                "document_count": document_count,
            },
        )

File: rust/operonx/src/core/exceptions.rs
//! Unified exception types for Operon.
//!
//! Mirrors Python `operonx/core/exceptions.py`:
//!
//! - [`OpError`] — op-level error hierarchy (Parser / Code / Branch / Condition /
//!   Iteration / Prompt / Embedding / Rerank), matching Python's `OpError` subclasses.
//!   Each variant carries the same structured context fields Python carries, so
//!   callers can pattern-match the kind AND extract diagnostic detail.
//! - [`OperonError`] — top-level error type wrapping `OpError` plus engine-level
//!   categories (Provider, ResourceHub, Config, State, Runtime).
//!
//! Python's subclass hierarchy collapses to Rust enum variants here.
//! `matches!(err, OperonError::Op(OpError::Parser { .. }))` replaces
//! `isinstance(err, ParserError)`.
//!
//! ## Display format
//!
//! `Display` (and therefore `to_string()`) produce Python's multi-line layout:
//!
//! ```text
//! [PARSER] Invalid JSON syntax
//!   Error: <original>
//!   format: 'json'
//!   input: '{"broken": '
//! ```
//!
//! Substring tests covering `[TAG]`, the message body, the `Error: ...` line,
//! and `key: value` entries — the same shape Python tests in
//! `tests/internal/core/test_exceptions.py` assert against — pass on both sides.

use std::collections::BTreeMap;
use std::error::Error as StdError;
use std::fmt;
use std::path::PathBuf;

/// Truncate a string with a Python-style `...` suffix. Mirrors
/// [`operonx.core.exceptions._truncate`](../../../../../operonx/core/exceptions.py).
pub fn truncate(text: &str, max_length: usize) -> String {
    if text.chars().count() <= max_length {
        text.to_string()
    } else {
        let prefix: String = text.chars().take(max_length).collect();
        format!("{prefix}...")
    }
}

/// Format a context value the way Python's `_format_message` does:
/// string values get `repr()`-style single quotes + `_truncate` truncation;
/// everything else uses Debug formatting. Best-effort parity — exact byte
/// equality with Python isn't required (tests use substring assertions).
fn fmt_context_value(value: &ContextValue) -> String {
    match value {
        ContextValue::Str(s) => {
            let q = format!("'{}'", s.replace('\'', "\\'"));
            truncate(&q, 200)
        }
        ContextValue::Int(n) => n.to_string(),
        ContextValue::Float(f) => f.to_string(),
        ContextValue::List(items) => {
            let inner: Vec<String> = items
                .iter()
                .map(|v| match v {
                    ContextValue::Str(s) => format!("'{s}'"),
                    other => fmt_context_value(other),
                })
                .collect();
            format!("[{}]", inner.join(", "))
        }
        ContextValue::Dict(m) => {
            let inner: Vec<String> = m
                .iter()
                .map(|(k, v)| format!("'{}': {}", k, fmt_context_value(v)))
                .collect();
            format!("{{{}}}", inner.join(", "))
        }
        ContextValue::Bool(b) => b.to_string(),
        ContextValue::Null => "None".into(),
    }
}

/// Typed context value for structured fields. Bridges to JSON Value for
/// round-tripping with the JSON wire layer; kept narrow so Display formatting
/// matches Python's `repr()` shape for the common cases.
#[derive(Debug, Clone)]
pub enum ContextValue {
    Str(String),
    Int(i64),
    Float(f64),
    Bool(bool),
    Null,
    List(Vec<ContextValue>),
    Dict(BTreeMap<String, ContextValue>),
}

impl From<&str> for ContextValue {
    fn from(s: &str) -> Self {
        ContextValue::Str(s.to_string())
    }
}
impl From<String> for ContextValue {
    fn from(s: String) -> Self {
        ContextValue::Str(s)
    }
}
impl From<i64> for ContextValue {
    fn from(n: i64) -> Self {
        ContextValue::Int(n)
    }
}
impl From<usize> for ContextValue {
    fn from(n: usize) -> Self {
        ContextValue::Int(n as i64)
    }
}
impl From<bool> for ContextValue {
    fn from(b: bool) -> Self {
        ContextValue::Bool(b)
    }
}
impl<T: Into<ContextValue>> From<Vec<T>> for ContextValue {
    fn from(v: Vec<T>) -> Self {
        ContextValue::List(v.into_iter().map(Into::into).collect())
    }
}

/// Optional "original error" carried alongside the op-error variant — the
/// upstream cause of the failure. `Display` formats as Python's
/// `f"  Error: {self.original_error}"` line.
pub type OriginalError = Option<String>;

/// Op-level error categories, mirroring Python's `OpError` subclass hierarchy.
///
/// Each variant carries the structured context fields the matching Python
/// subclass takes, in the same names. Use `matches!(e, OpError::Parser { .. })`
/// to pattern-match by category; `if let OpError::Parser { input_text, format_type, .. } = e`
/// to extract specific fields.
#[derive(Debug, Clone)]
pub enum OpError {
    /// Mirrors Python `ParserError(message, input_text, format_type, original_error)`.
    Parser {
        message: String,
        input_text: String,
        format_type: String,
        original_error: OriginalError,
    },
    /// Mirrors Python `CodeError(message, function_name, source, inputs, original_error)`.
    Code {
        message: String,
        function_name: String,
        source: String,
        inputs: BTreeMap<String, String>,
        original_error: OriginalError,
    },
    /// Mirrors Python `BranchError(message, condition, inputs, candidates, original_error)`.
    Branch {
        message: String,
        condition: String,
        inputs: BTreeMap<String, String>,
        candidates: Vec<String>,
        original_error: OriginalError,
    },
    /// Mirrors Python `ConditionError(message, condition, inputs, iteration, phase, original_error)`.
    Condition {
        message: String,
        condition: String,
        inputs: BTreeMap<String, String>,
        iteration: Option<usize>,
        phase: String,
        original_error: OriginalError,
    },
    /// Mirrors Python `IterationError(message, iteration_index, loop_data, total_iterations, op_type, original_error)`.
    Iteration {
        message: String,
        iteration_index: usize,
        loop_data: BTreeMap<String, String>,
        total_iterations: usize,
        op_type: String,
        original_error: OriginalError,
    },
    /// Mirrors Python `PromptError(message, template, missing_vars, original_error)`.
    Prompt {
        message: String,
        template_type: String,
        template: String,
        missing_vars: Vec<String>,
        original_error: OriginalError,
    },
    /// Mirrors Python `EmbeddingError(message, resource, text_count, original_error)`.
    Embedding {
        message: String,
        resource: String,
        text_count: usize,
        original_error: OriginalError,
    },
    /// Mirrors Python `RerankError(message, resource, query, document_count, original_error)`.
    Rerank {
        message: String,
        resource: String,
        query: String,
        document_count: usize,
        original_error: OriginalError,
    },
}

impl OpError {
    /// Short tag matching Python's `op_type.upper()` — used as the `[TAG]`
    /// prefix in `Display` and as a discriminator for parity tests.
    pub fn tag(&self) -> &'static str {
        match self {
            OpError::Parser { .. } => "PARSER",
            OpError::Code { .. } => "CODE",
            OpError::Branch { .. } => "BRANCH",
            OpError::Condition { .. } => "WHILE",
            OpError::Iteration { .. } => "FOR",
            OpError::Prompt { .. } => "PROMPT",
            OpError::Embedding { .. } => "EMBEDDING",
            OpError::Rerank { .. } => "RERANK",
        }
    }

    /// Short lowercase kind matching Python's `op_type` field — useful for
    /// structured emission to telemetry / parity-fixture comparison.
    pub fn kind(&self) -> &'static str {
        match self {
            OpError::Parser { .. } => "parser",
            OpError::Code { .. } => "code",
            OpError::Branch { .. } => "branch",
            OpError::Condition { .. } => "while",
            OpError::Iteration { .. } => "for",
            OpError::Prompt { .. } => "prompt",
            OpError::Embedding { .. } => "embedding",
            OpError::Rerank { .. } => "rerank",
        }
    }

    /// User-supplied message (the first positional arg every constructor takes).
    pub fn message(&self) -> &str {
        match self {
            OpError::Parser { message, .. }
            | OpError::Code { message, .. }
            | OpError::Branch { message, .. }
            | OpError::Condition { message, .. }
            | OpError::Iteration { message, .. }
            | OpError::Prompt { message, .. }
            | OpError::Embedding { message, .. }
            | OpError::Rerank { message, .. } => message,
        }
    }

    /// Original upstream error, if one was attached.
    pub fn original_error(&self) -> Option<&str> {
        let oe = match self {
            OpError::Parser { original_error, .. }
            | OpError::Code { original_error, .. }
            | OpError::Branch { original_error, .. }
            | OpError::Condition { original_error, .. }
            | OpError::Iteration { original_error, .. }
            | OpError::Prompt { original_error, .. }
            | OpError::Embedding { original_error, .. }
            | OpError::Rerank { original_error, .. } => original_error,
        };
        oe.as_deref()
    }

    /// Variant-specific context as `(key, formatted-value)` pairs, in the same
    /// order Python emits them in `_format_message`. Returned as borrowed
    /// strings to avoid allocation in the no-context path.
    fn context_lines(&self) -> Vec<(String, String)> {
        match self {
            OpError::Parser {
                input_text,
                format_type,
                ..
            } => vec![
                (
                    "format".into(),
                    fmt_context_value(&ContextValue::Str(format_type.clone())),
                ),
                (
                    "input".into(),
                    fmt_context_value(&ContextValue::Str(input_text.clone())),
                ),
            ],
            OpError::Code {
                function_name,
                source,
                inputs,
                ..
            } => {
                let inputs_map: BTreeMap<String, ContextValue> = inputs
                    .iter()
                    .map(|(k, v)| (k.clone(), ContextValue::Str(truncate(v, 100))))
                    .collect();
                vec![
                    (
                        "function_name".into(),
                        fmt_context_value(&ContextValue::Str(function_name.clone())),
                    ),
                    (
                        "source".into(),
                        fmt_context_value(&ContextValue::Str(truncate(source, 300))),
                    ),
                    (
                        "inputs".into(),
                        fmt_context_value(&ContextValue::Dict(inputs_map)),
                    ),
                ]
            }
            OpError::Branch {
                condition,
                inputs,
                candidates,
                ..
            } => {
                let inputs_map: BTreeMap<String, ContextValue> = inputs
                    .iter()
                    .map(|(k, v)| (k.clone(), ContextValue::Str(truncate(v, 100))))
                    .collect();
                vec![
                    (
                        "condition".into(),
                        fmt_context_value(&ContextValue::Str(condition.clone())),
                    ),
                    (
                        "inputs".into(),
                        fmt_context_value(&ContextValue::Dict(inputs_map)),
                    ),
                    (
                        "candidates".into(),
                        fmt_context_value(&ContextValue::List(
                            candidates
                                .iter()
                                .map(|c| ContextValue::Str(c.clone()))
                                .collect(),
                        )),
                    ),
                ]
            }
            OpError::Condition {
                condition,
                inputs,
                iteration,
                phase,
                ..
            } => {
                let mut out = vec![
                    (
                        "condition".into(),
                        fmt_context_value(&ContextValue::Str(condition.clone())),
                    ),
                    (
                        "phase".into(),
                        fmt_context_value(&ContextValue::Str(phase.clone())),
                    ),
                ];
                if !inputs.is_empty() {
                    let inputs_map: BTreeMap<String, ContextValue> = inputs
                        .iter()
                        .map(|(k, v)| (k.clone(), ContextValue::Str(truncate(v, 100))))
                        .collect();
                    out.push((
                        "inputs".into(),
                        fmt_context_value(&ContextValue::Dict(inputs_map)),
                    ));
                }
                if let Some(i) = iteration {
                    out.push(("iteration".into(), i.to_string()));
                }
                out
            }
            OpError::Iteration {
                iteration_index,
                loop_data,
                total_iterations,
                ..
            } => {
                let data_map: BTreeMap<String, ContextValue> = loop_data
                    .iter()
                    .map(|(k, v)| (k.clone(), ContextValue::Str(truncate(v, 100))))
                    .collect();
                vec![
                    (
                        "iteration_index".into(),
                        format!("{}/{}", iteration_index, total_iterations),
                    ),
                    (
                        "loop_data".into(),
                        fmt_context_value(&ContextValue::Dict(data_map)),
                    ),
                ]
            }
            OpError::Prompt {
                template_type,
                template,
                missing_vars,
                ..
            } => {
                let mut out = vec![
                    (
                        "template_type".into(),
                        fmt_context_value(&ContextValue::Str(template_type.clone())),
                    ),
                    (
                        "template".into(),
                        fmt_context_value(&ContextValue::Str(truncate(template, 300))),
                    ),
                ];
                if !missing_vars.is_empty() {
                    out.push((
                        "missing_vars".into(),
                        fmt_context_value(&ContextValue::List(
                            missing_vars
                                .iter()
                                .map(|v| ContextValue::Str(v.clone()))
                                .collect(),
                        )),
                    ));
                }
                out
            }
            OpError::Embedding {
                resource,
                text_count,
                ..
            } => vec![
                (
                    "resource".into(),
                    fmt_context_value(&ContextValue::Str(resource.clone())),
                ),
                ("text_count".into(), text_count.to_string()),
            ],
            OpError::Rerank {
                resource,
                query,
                document_count,
                ..
            } => vec![
                (
                    "resource".into(),
                    fmt_context_value(&ContextValue::Str(resource.clone())),
                ),
                (
                    "query".into(),
                    fmt_context_value(&ContextValue::Str(truncate(query, 100))),
                ),
                ("document_count".into(), document_count.to_string()),
            ],
        }
    }
}

impl fmt::Display for OpError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "[{}] {}", self.tag(), self.message())?;
        if let Some(orig) = self.original_error() {
            writeln!(f, "  Error: {}", orig)?;
        }
        let lines = self.context_lines();
        for (i, (k, v)) in lines.iter().enumerate() {
            if i == lines.len() - 1 {
                write!(f, "  {}: {}", k, v)?;
            } else {
                writeln!(f, "  {}: {}", k, v)?;
            }
        }
        Ok(())
    }
}

impl StdError for OpError {}

// ── Convenience constructors — terse `OpError::code("msg", "fn", ...)` style ──

impl OpError {
    /// Short-form constructor: fill optional structured fields with defaults.
    /// Mirrors the common case where a caller has just a message + the op's
    /// own name. For richer context use the struct literal form directly.
    pub fn code_msg(message: impl Into<String>, function_name: impl Into<String>) -> Self {
        OpError::Code {
            message: message.into(),
            function_name: function_name.into(),
            source: String::new(),
            inputs: BTreeMap::new(),
            original_error: None,
        }
    }

    pub fn parser_msg(message: impl Into<String>, format_type: impl Into<String>) -> Self {
        OpError::Parser {
            message: message.into(),
            input_text: String::new(),
            format_type: format_type.into(),
            original_error: None,
        }
    }

    pub fn branch_msg(message: impl Into<String>, condition: impl Into<String>) -> Self {
        OpError::Branch {
            message: message.into(),
            condition: condition.into(),
            inputs: BTreeMap::new(),
            candidates: Vec::new(),
            original_error: None,
        }
    }

    pub fn prompt_msg(message: impl Into<String>) -> Self {
        OpError::Prompt {
            message: message.into(),
            template_type: String::new(),
            template: String::new(),
            missing_vars: Vec::new(),
            original_error: None,
        }
    }

    pub fn embedding_msg(
        message: impl Into<String>,
        resource: impl Into<String>,
        text_count: usize,
    ) -> Self {
        OpError::Embedding {
            message: message.into(),
            resource: resource.into(),
            text_count,
            original_error: None,
        }
    }

    pub fn rerank_msg(
        message: impl Into<String>,
        resource: impl Into<String>,
        document_count: usize,
    ) -> Self {
        OpError::Rerank {
            message: message.into(),
            resource: resource.into(),
            query: String::new(),
            document_count,
            original_error: None,
        }
    }
}

/// Top-level error type for all Operon operations.
///
/// Wraps [`OpError`] for op-level failures and adds engine-level categories.
#[derive(Debug, Clone, thiserror::Error)]
pub enum OperonError {
    #[error("{0}")]
    Op(#[from] OpError),

    #[error("provider error: {0}")]
    Provider(String),

    #[error("resource hub: {0}")]
    ResourceHub(String),

    #[error("config: {0}")]
    Config(String),

    #[error("state: {0}")]
    State(String),

    #[error("runtime: {0}")]
    Runtime(String),

    #[error("schema: unsupported schema_version {0}; expected {expected}", expected = SUPPORTED_SCHEMA_VERSION)]
    UnsupportedSchema(String),

    /// `${VAR}` reference in `resources.yaml` could not be resolved.
    ///
    /// Distinct from [`OperonError::ResourceHub`] / [`OperonError::Config`] so
    /// callers can match specifically on missing-env-var setup problems and
    /// give the user a tailored remediation hint. Mirrors Python's
    /// `EnvVarUnsetError` (subclass of `RuntimeError`).
    #[error(
        "resource '{key}' references unset environment variable {var:?}\n  source: {source_path:?}\n  .env paths searched: {env_paths:?}"
    )]
    EnvVarUnset {
        var: String,
        key: String,
        source_path: Option<PathBuf>,
        env_paths: Vec<PathBuf>,
    },
}

/// The serialized graph schema version Rust accepts. Bump on breaking changes.
pub const SUPPORTED_SCHEMA_VERSION: &str = "1.0";

// ── From impls for common upstream errors ─────────────────────────────────

impl From<serde_json::Error> for OperonError {
    fn from(e: serde_json::Error) -> Self {
        OperonError::Config(e.to_string())
    }
}

impl From<serde_yaml::Error> for OperonError {
    fn from(e: serde_yaml::Error) -> Self {
        OperonError::Config(e.to_string())
    }
}

impl From<std::io::Error> for OperonError {
    fn from(e: std::io::Error) -> Self {
        OperonError::Runtime(e.to_string())
    }
}

/// Convenience alias for `Result<T, OperonError>`.
pub type Result<T> = std::result::Result<T, OperonError>;

/// Boxed dynamic error for cases where the source type is opaque (e.g., provider HTTP errors).
pub type BoxError = Box<dyn StdError + Send + Sync + 'static>;
